rfd/0000-prototype-job-worker.md

```diff
@@ -0,0 +1,203 @@
+  handling all server responses and streaming outputs effectively.
+
+### Detailed Design
+#### Streaming Approach
+- **Implementation**: Utilize gRPC streaming to send output from the worker to
+  the client in real-time. Buffering strategies will be applied to handle
+  high-frequency output efficiently.
+#### TLS Setup
+- **TLS Version**: TLS 1.3 will be used due to its enhanced security features.
+- **Cipher Suites**: Use strong cipher suites such as TLS_AES_256_GCM_SHA384 to
+  ensure secure communication.
+- **Certificate Management**: Use hardcoded certificates for simplicity, with
+  TODO items added for future implementation of a more robust certificate
+  management system.
+#### Process Execution Lifecycle
+1. **Start Job**:
+    - Jobs are initiated via a `StartJob` RPC call, which includes the command and its arguments.
+    - Each job is assigned to a unique cgroup for resource management.
+    - Jobs are executed within isolated namespaces to prevent interference and enhance security.
+2. **Stream Output**:
+    - Output from the process is captured from the start and streamed to the
+      client using gRPC's streaming capabilities. 
+3. **Stop Job**:
+    - Jobs can be terminated using a StopJob RPC call.
+    - Processes are immediately killed, ensuring all resources are properly released.
+4. **Query Status**:
+    - The status of any job can be queried to return whether it is running,
+      stopped, or completed. For this initial prototype, the information provided
+      will be basic. However, a final product would expand on this to include
+      more detailed information, such as the return code, error output, and other
+      relevant diagnostics to provide a comprehensive view of the job's execution
+      status.
+
+#### Namespace Implementation
+Namespaces are a feature of the Linux kernel that provides isolation for
+running processes. They make it possible for processes to have their own view
+of the system's global resources, such as process IDs, network stacks, mount
+points, and interprocess communication resources. We will specifically use the
+following namespaces:
+  - **PID Namespace**: Ensures process isolation by providing
+    processes with separate PID namespaces. This prevents them from seeing or
+    interacting with processes in other namespaces.
+  - **Network Namespace**: Isolates network interfaces, IP routing tables, port
+    numbers, etc., allowing processes to operate on a network as if they were
+    the only ones on the system. To stay in scope, the network will remain
+    isolated.
+  - **Mount Namespace**: Isolates filesystem mount points, allowing processes
+    to have their own set of filesystems visible to them.
+
+##### Utilization Strategy
+Our service will utilize the syscall package in Go to unshare and manage
+namespaces. The following code snippet demonstrates how a new process can be
+started in a new set of namespaces:
```

Comment:
How a new process will be started ? 

and when `setupNamespaces` will be called ?

Reply:
See https://github.com/bill-rich/jobworker/pull/1#discussion_r1609012189